controllers/author-controller.js

```diff
@@ -0,0 +1,210 @@
+var Book = require("../models/book");
+
+var async = require("async");
+
+var { body, validationResult } = require("express-validator");
+
+// Display list of all Authors.
+exports.author_list = function (req, res) {
+  Author.find()
+    .sort([["family_name", "ascending"]])
+    .exec(function (err, list_authors) {
+      if (err) {
+        return next(err);
+      }
+      //Successful, so render
+      res.render("author-list", {
+        title: "Author List",
+        author_list: list_authors,
+      });
+    });
+};
+
+// Display detail page for a specific Author.
+exports.author_detail = function (req, res) {
+  async.parallel(
+    {
+      author: function (callback) {
+        Author.findById(req.params.id).exec(callback);
+      },
+      authors_books: function (callback) {
+        Book.find({ author: req.params.id }, "title summary").exec(callback);
+      },
+    },
+    function (err, results) {
+      if (err) {
+        return next(err);
+      } // Error in API usage.
+      if (results.author == null) {
+        // No results.
+        var err = new Error("Author not found");
+        err.status = 404;
+        return next(err);
+      }
+      // Successful, so render.
+      res.render("author-detail", {
+        title: "Author Detail",
+        author: results.author,
+        author_books: results.authors_books,
+      });
+    }
+  );
+};
+
+// Display Author create form on GET.
+exports.author_create_get = function (req, res) {
+  res.render("author-form", { title: "Create Author" });
+};
+
+// Handle Author create on POST.
+exports.author_create_post = [
+  // Validate and sanitise fields.
+  body("first_name")
+    .trim()
+    .isLength({ min: 2 })
+    .escape()
+    .withMessage("First name must be specified.")
+    .isAlphanumeric()
+    .withMessage("First name has non-alphanumeric characters."),
+  body("family_name")
+    .trim()
+    .isLength({ min: 2 })
+    .escape()
+    .withMessage("Family name must be specified.")
+    .isAlphanumeric()
+    .withMessage("Family name has non-alphanumeric characters."),
+  body("date_of_birth", "Invalid date of birth")
+    .optional({ checkFalsy: true })
+    .isISO8601()
+    .toDate(),
+  body("date_of_death", "Invalid date of death")
+    .optional({ checkFalsy: true })
+    .isISO8601()
+    .toDate(),
+
+  // Process request after validation and sanitization.
+  (req, res, next) => {
+    // Extract the validation errors from a request.
+    const errors = validationResult(req);
+
+    if (!errors.isEmpty()) {
+      // There are errors. Render form again with sanitized values/errors messages.
+      res.render("author-form", {
+        title: "Create Author",
+        author: req.body,
+        errors: errors.array(),
+      });
+      return;
+    } else {
+      // Data from form is valid.
+
+      // Create an Author object with escaped and trimmed data.
+      var author = new Author({
+        first_name: req.body.first_name,
```

Comment:
Please extract and store values at the beginning of the function.

Reply:
Please make these changes through out from now for all params, query and body of the req.

Reply:
already done sir